### ISSUE TITLE
Login 401 from Postman client

### ISSUE THREAD

**Issue by Emixam23** [mangledbottles/Musically-API#7]
Hey,

I was trying to connect to musycal.ly from a postman client so I tried to understand the Php code of [src/musically.php](https://github.com/mangledbottles/Musically-API/tree/master/src), however, I am getting a 401 but I ain't sure about why.

There is my request:

**[POST]** https://api.musical.ly/rest/passport/v2/login?supportLoginVerify=true

**Headers** *(22)*:

![screen shot 2018-03-19 at 10 47 01](https://user-images.githubusercontent.com/6987848/37602528-e7ea7d88-2b62-11e8-8159-cd4cecff1dd9.png)

**x-www-form-encoded**:

<img width="541" alt="url form encoded" src="https://user-images.githubusercontent.com/6987848/37603003-fcf21c76-2b63-11e8-8c75-a099700d951f.png">

And I am getting:

    {
        "timestamp": 1521470164701,
        "status": 401,
        "error": "Unauthorized",
        "exception": "net.vickymedia.passport.composite.exception.UnauthorizedException",
        "message": "Unauthorized",
        "path": "/v2/login"
    }

Do you have any idea why? The 3 Variables **USERNAME**, **USER_EMAIL**, and **USER_PASSWORD** are replaced by the correct values of course.

Thanks for any update !

**Comment by charlieAndroidDev** [comment]
Most likely is that your X-Request-Sign5 header value is invalid.  The header value is a hash of the request info which contains a time stamp, if the time stamp is very old, like when this library was written then the server will most likely say you are unauthorised.  You need to generate your own X-Request-Info5 header value and then your own X-Request-Sign5 header value to get the correct response.

**Comment by Emixam23** [comment]
Hey, thanks for your answer !

So I just need to generate a HMAC-SHA1 hash of a current timestamp?

*You need to generate your own X-Request-Info5 header value* What is it corresponding to?

Thanks !

**Comment by charlieAndroidDev** [comment]
Base64 decode your current X-Request-Info5 header value, replace the time stamp with a current one and then encode it with base64 again.  Generate an HMAC-SHA1 hash of that new value and that is the new X-Request-Sign5.

**Comment by Emixam23** [comment]
Okay thank you, last question, do you know by any chance the key to use for the HMAC-SHA1 hash?

**Comment by charlieAndroidDev** [comment]
It depends on whether or not you want to emulate the Android app or the IOS app.

**Comment by Emixam23** [comment]
Since in the header I am giving the information that I am using an iOS device, let's use the full iOS logic?

**Comment by charlieAndroidDev** [comment]
Here's a helpful article I wrote for everyone: https://medium.com/@charliethinks/reverse-engineering-musical-y-live-ly-android-apps-part-1-a910daad2ec2

**Comment by Emixam23** [comment]
Hey !

I read a bit your article, really interesting ! Thank you :)

However, I am getting stuck, not at the article, but the data itself.

At a moment, you're talking about those data:

    {
        “os”:”android 7.0",
        ”version”:”6.0.10",
        ”slider-show-cookie”:””,
        ”X-Request-ID”:”56aa90d1–3c01–4310–9dbf-2db48c676a3c”,
        ”method”:”GET”,
        ”url”:”https:\/\/live.direct.ly\/rest\/discover\/navigate\/nonlogin?timestamp=1520746145851&”,
        ”ostype”:”android”,
        ”deviceid”:”a014ca20202f8e48169002d1c2ad74620b5499",
        ”timestamp”:1520746145873
    }

And I was wondering how can you achieve the only backslash in the middle of the string? I feel like I didn't get it, since when can you put only one backslash like that? we all the time need 2 since the backslash alone is an option for formatting?

Thanks in adavance

**Comment by Emixam23** [comment]
Hey, nice link, really interesting !

I was just wondering if you could explain to me the "M-TOKEN hash=" logic please?
Also, I don't know if the signature works X-Request* works since my token is probably wrong :/

But thanks !

**Comment by charlieAndroidDev** [comment]
The "M-TOKEN hash=" is given to you as a header after login I believe, or at least a very specific endpoint.

**Comment by Emixam23** [comment]
Ho yeah, that's why I thought and you're right, you're getting it at the login, thanks to `Packet Capture` ! 

In order to login, you need to use the link of the discover nonlogin?

**Comment by Emixam23** [comment]
Hey,

Yeah it comes from the login response, thanks.

Max

**Comment by Emixam23** [comment]
Success (y) Thanks for your help

**Comment by Emixam23** [comment]
@mangledbottles 

If were you, I would wait just a bit, I am currently writing an updated version. Once I am done, I will send it to @charlieAndroidDev .

After then, he will be able to make maybe a more understandable code.

Best,

**Comment by charlieAndroidDev** [comment]
Yeah I’m happy to do that.  There are gonna be pretty huge changes though...

**Comment by mangledbottles** [comment]
@charlieAndroidDev has it changed that much?
They didn't seem to care too much about checking security precautions when I first looked at it.

**Comment by Emixam23** [comment]
@mangledbottles Not that much, but it changed

**Comment by NullUsers** [comment]
@charlieAndroidDev Thank you 👏👏

**Comment by AeonLucid** [comment]
@charlieAndroidDev Woah, I should have found your medium write up earlier. I took a different approach to get the signature.. which works but.. it's not easy to setup for consistent use. See my repository (which also includes a c# musical.ly API): https://github.com/AeonLucid/MusicallyRE

**Comment by charlieAndroidDev** [comment]
@farrusete The IOS app may well use a different secret key to the one used in the Android app.  I have not reverse engineered the IOS app whatsoever and as such cannot comment on whether replicating an iOS request in such a way would work at all.
